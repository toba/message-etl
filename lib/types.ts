export enum Source {
   FacebookMessenger,
   GoogleVoice,
   GMail,
   MicrosoftChat,
   Passport
}

export enum Relation {
   None = 0,
   Self,
   Other
}

export interface Message {
   /** Who sent the message. */
   from: Relation;
   /** When the message was sent. */
   on: Date;
   subject?: string;
   /** The platform used to convey the message. */
   source: Source;
   /** Content of the message. */
   text: string | null;
   /** An emoji reaction. */
   reaction?: string;
}

/**
 * Methods to process and parse messages from a particular platform.
 */
export interface Reader {
   sourceName: string;
   /** Filter file list so it only includes those for the platform. */
   filter: (fileName: string) => boolean;
   /** Process an exported file. */
   process: (fileText: string, fileName: string) => Message[];
   /** Parse a single message from the exported file. */
   parse: (msg: any) => Message | null;
}

export interface Writer {
   serialize: (m: Message) => string;
   write: (messages: Message[]) => string;
   fileName: string;
}

export interface Person {
   name: string;
   alias: string[];
   email: string[];
   phone: string[];
}

export type People = { [key: number]: Person };

export namespace Chat {
   export interface User {
      FriendlyName: string;
      LogonName: string;
   }

   export interface Participant {
      User: User;
   }

   export interface Text {
      '#text': string;
      Style: string;
   }

   export interface Message {
      SessionID: number;
      Time: string;
      Date: string;
      DateTime: string;
      From: Participant;
      To: Participant;
      Text: string;
   }

   export interface Log {
      FirstSessionID: number;
      LastSessionID: number;
      LogonName: string;
      Message: Message[];
   }

   export interface Conversation {
      Log: Log;
   }
}

export namespace Facebook {
   export interface Reaction {
      /** Emoji */
      reaction: string;
      /** Person's name */
      actor: string;
   }

   export interface Link {
      uri: string;
   }

   export interface MediaLink extends Link {
      creation_timestamp: number;
   }

   export interface VideoLink extends MediaLink {
      thumbnail: Link;
   }

   export interface Message {
      sender_name: string;
      timestamp_ms: number;
      content: string;
      type: 'Generic';
      sticker?: Link;
      videos?: VideoLink[];
      photos?: MediaLink[];
      reactions?: Reaction[];
   }

   interface Participant {
      name: string;
   }

   /**
    * Exported Facebook Messenger data. Export files are per-conversation.
    * @see https://www.facebook.com/help/1701730696756992?helpref=hc_global_nav
    */
   export interface Conversation {
      participants: Participant[];
      messages: Message[];
      title: string;
      is_still_participant: boolean;
      thread_type: string;
      thread_path: string;
   }
}

export namespace Passport {
   export interface Message {
      /** Content */
      RD: string;
      /** From */
      RM: string;
      /** To */
      RO: string;
      /** Time */
      RT: string;
      /** Added during processing */
      date?: string;
      /** Added during processing */
      defaultFrom?: string;
   }

   /** Misspellings in original */
   export interface Conversation {
      ContactFirendName: string;
      ContactLoginName: string;
      CreateDate: string;
      OwnerFirendName: string;
      OwnerLoginName: string;
      RS: Message[];
   }

   export interface History {
      History: Conversation[];
   }

   export interface Log {
      Historys: History;
   }
}
