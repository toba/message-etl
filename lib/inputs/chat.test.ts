import '@toba/test';
import { readFileText } from '@toba/test';
import { chat } from './index';
import { Relation } from '../types';

jest.mock('../config/index');

test('matches file names', () => {
   [
      'other 2003-09-11 home.xml',
      'other 2004-06-15 work.xml',
      'other3008@hotmail.com.xml',
      'other30083257279302.xml'
   ].forEach(n => {
      expect(chat.filter(n)).toBe(true);
   });

   [
      'Family-Name.mbox',
      'message.json',
      'Other - Voicemail - 2010-10-30T01_.mp3',
      'other 2003-06-05 work.xml',
      'other 2003-08-13 home.xml'
   ].forEach(n => {
      expect(chat.filter(n)).toBe(false);
   });
});

test('parses file', async () => {
   const text = await readFileText(`${__dirname}/__mocks__/chat.xml`);
   const messages = chat.process(text, 'mock');

   expect(messages).toBeInstanceOf(Array);
   expect(messages).toHaveLength(5);
   expect(messages[0].from).toBe(Relation.Other);
   expect(messages[1].from).toBe(Relation.Self);
   expect(messages[3].text).toBe('sorry, been on and still on the phone');

   expect(messages).toMatchSnapshot();
});
