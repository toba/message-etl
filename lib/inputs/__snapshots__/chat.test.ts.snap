// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses file 1`] = `
Array [
  Object {
    "from": 2,
    "on": 2003-06-24T16:24:17.396Z,
    "source": 3,
    "text": "let me guess, working hard again?",
  },
  Object {
    "from": 1,
    "on": 2003-06-24T16:24:42.102Z,
    "source": 3,
    "text": "yes",
  },
  Object {
    "from": 2,
    "on": 2003-06-24T16:25:54.216Z,
    "source": 3,
    "text": "Laura wearing you out yet?",
  },
  Object {
    "from": 1,
    "on": 2003-06-24T16:29:06.953Z,
    "source": 3,
    "text": "sorry, been on and still on the phone",
  },
]
`;
